### PR TITLE
Fix rb_interned_str_* functions to not assume static strings

### PR DESCRIPTION
**Pull request by casperisfine**
Followup: https://github.com/ruby/ruby/pull/3586

When passed a `fake_str`, `register_fstring` would create new strings with `str_new_static`. That's not what was expected, and answer almost no use cases.

### CODE REVIEW

**Comment by XrXr** [comment]
It makes sense to me that these functions should work on buffers that are not C string literals.

In that case, we are creating some new APIs instead of merely exposing some previously internal one without modification,  meaning the title of the [Redmine ticket](https://bugs.ruby-lang.org/issues/13381) is kind of a misnomer. I being this up to make sure that we are on the same page about what these new functions are supposed to do. 

Code wise,  I would prefer passing an additional parameter to `fstr_update_callback` instead of copy pasting and modifying it.

File: ext/-test-/string/rb_interned_str.c [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#include "ruby.h"
+
+static VALUE
+bug_rb_interned_str_dup(VALUE self, VALUE str)
+{
+    rb_check_type(str, T_STRING);
```
Comment (XrXr) [inline ext/-test-/string/rb_interned_str.c:6]:
```suggestion
    Check_Type(str, T_STRING);
```
`Check_Type` is more common.

**Comment by casperisfine** [comment]
> Code wise, I would prefer passing an additional parameter to `fstr_update_callback` instead of copy pasting and modifying it.

Yes that a good point. I'll try to clean this up.

**Comment by casperisfine** [comment]
@ko1 I'm sorry to bother you again with this but could you have a look? Without this PR I don't think `rb_interned_str*` is usable.

**Comment by ko1** [comment]
Thank you I'm asking @nobu

**Review by nobu — APPROVED** [review]

**Comment by casperisfine** [comment]
Thanks @nobu !

**Comment by casperisfine** [comment]
@ko1 is this good for you?

**Comment by ko1** [comment]
Ah, I have no opinion, thank you so much to provide your efforts!

**Comment by casperisfine** [comment]
Thanks!

**Pull request merged**